core/src/storage.cpp

```diff
@@ -217,6 +221,22 @@ void SolutionSequence::fillMessage(moveit_task_constructor_msgs::Solution& msg,
 	}
 }
 
+double SolutionSequence::computeCost(const CostTerm& f, std::string& comment) const {
+	double cost{ 0.0 };
+	std::string subcomment;
+	for (auto& solution : subsolutions_) {
+		cost += solution->computeCost(f, subcomment);
+		if (!subcomment.empty()) {
```

Comment:
> Do you really want to aggregate all comments? This will become unreadable for long sequences anyway.

Yes and mostly yes. I still don't like the use-case to specify such a cost function for big containers anyway, but we do want to be able to provide that information as well.

> Considering, Clearance as an example, one would like to set a single comment mentioning the accumulated clearance across all sub solutions, right?

My primary reason for adding the comment for Clearance was not to report the accumulated clearance but
- to get feedback in case of detected collisions and
- to see the names of the two links that have the reported distance for InterfaceState checks

It might make sense to generalize the comment for estimating trajectories to include all relevant link pairs there as well.

> This makes me think, that we should shift the hierarchical traversal/visitor pattern from the Solution class(es) to the CostTerm class. This would allow CostTerms to override the default accumulation behavior for sequences.

This relates to https://github.com/ros-planning/moveit_task_constructor/pull/183#discussion_r486250809 .

I wanted to keep CostTerm altogether optional because I consider them an extension. So I did not put any core logic into the class. The more we tangle the structures/classes, the harder the core sources are to understand and we keep getting feedback that things are way too complicated already.

Reply:
> I wanted to keep CostTerm altogether optional because I consider them an extension.

I don't see how this comment relates to mine. Neither is CostTerm optional - it's the basic class used for costs. Yes, you can initialize a CostTerm with a simple lambda function, but this doesn't stop you from implementing the visitor pattern in this class, does it?

Reply:
> I don't see how this comment relates to mine.

Nor do I two days later. Sorry I should have thought about it more before replying.

However, [the corresponding commit](https://github.com/ros-planning/moveit_task_constructor/pull/183/commits/be93501136fb59ba10e07f7cb46e736e0145c0bb) gives quite a lot of reasoning about why I went for the support flag approach. You might have missed it?

However, with the aggregator removed again, half the reasoning does not apply anymore, so I implemented the vtable approach now.